p2p/protocol/identify/obsaddr.go

```diff
@@ -52,40 +53,51 @@ func (oa *ObservedAddr) activated(ttl time.Duration) bool {
 }
 
 type newObservation struct {
-	observed, local, observer ma.Multiaddr
-	direction                 network.Direction
+	conn     network.Conn
+	observed ma.Multiaddr
 }
 
-// ObservedAddrSet keeps track of a set of ObservedAddrs
-// the zero-value is ready to be used.
-type ObservedAddrSet struct {
-	sync.RWMutex // guards whole datastruct.
+// ObservedAddrManager keeps track of a ObservedAddrs.
+type ObservedAddrManager struct {
+	host host.Host
 
+	// latest observation from active connections
+	// we'll "re-observe" these when we gc
+	activeConnsMu sync.Mutex
+	// active connection -> most recent observation
+	activeConns map[network.Conn]ma.Multiaddr
+
+	mu sync.RWMutex
 	// local(internal) address -> list of observed(external) addresses
-	addrs map[string][]*ObservedAddr
-	ttl   time.Duration
+	addrs        map[string][]*ObservedAddr
+	ttl          time.Duration
+	refreshTimer *time.Timer
 
 	// this is the worker channel
 	wch chan newObservation
 }
 
-// NewObservedAddrSet returns a new set using peerstore.OwnObservedAddressTTL
-// as the TTL.
-func NewObservedAddrSet(ctx context.Context) *ObservedAddrSet {
-	oas := &ObservedAddrSet{
-		addrs: make(map[string][]*ObservedAddr),
-		ttl:   peerstore.OwnObservedAddrTTL,
-		wch:   make(chan newObservation, observedAddrSetWorkerChannelSize),
+// NewObservedAddrManager returns a new address manager using
+// peerstore.OwnObservedAddressTTL as the TTL.
+func NewObservedAddrManager(host host.Host) *ObservedAddrManager {
+	oas := &ObservedAddrManager{
+		addrs:        make(map[string][]*ObservedAddr),
+		ttl:          peerstore.OwnObservedAddrTTL,
+		wch:          make(chan newObservation, observedAddrManagerWorkerChannelSize),
+		host:         host,
+		activeConns:  make(map[network.Conn]ma.Multiaddr),
+		refreshTimer: time.NewTimer(peerstore.OwnObservedAddrTTL / 2),
```

Comment:
why `/2`?

Reply:
I'll add a comment. If we wait until the address TTL expires, the addresses will expire. So we wait half the expiration period before refreshing them.